zerver/lib/user_groups.py

```diff
@@ -29,6 +29,24 @@ class UserGroupDict(TypedDict):
     can_mention_group_id: int
 
 
+def user_ids_to_users(user_ids: Sequence[int], realm: Realm) -> List[UserProfile]:
+    # TODO: Consider adding a flag to control whether deactivated
+    # users should be included.
+
+    user_profiles = list(UserProfile.objects.filter(id__in=user_ids).select_related("realm"))
+
+    found_user_ids = {user_profile.id for user_profile in user_profiles}
+
+    for user_id in user_ids:
+        if user_id not in found_user_ids:
+            raise JsonableError(_("Invalid user ID: {user_id}").format(user_id=user_id))
+
+    for user_profile in user_profiles:
+        if user_profile.realm_id != realm.id:
+            raise JsonableError(_("Invalid user ID: {user_id}").format(user_id=user_profile.id))
```

Comment:
I also noticed that now that we don't have the cache, this block can be deleted in favor of just adding a filter above.